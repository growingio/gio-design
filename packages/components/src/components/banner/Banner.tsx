import React, { useState } from 'react';
import { Close } from '@gio-design/icons';
import className from 'classnames';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import { BannerProps } from './interface';

const Banner: React.FC<BannerProps> = (props: BannerProps) => {
  const { type = 'normal', content, closeable = true, onClose, button, prefixCls: customizePrefixCls } = props;
  const [visible, setVisible] = useState(true);
  const prefixCls = usePrefixCls('banner', customizePrefixCls);
  const onCloseBanner = () => {
    setVisible(false);
    onClose?.();
  };

  return (
    <div
      className={className(
        `${prefixCls}`,
        `${prefixCls}-${type}`,
        closeable && `${prefixCls}-closeable`,
        !visible && `${prefixCls}-close`
      )}
    >
      <div className={className(`${prefixCls}-content`, button && `${prefixCls}-content-button`)}>{content}</div>
      <div className={className(`${prefixCls}-button`)}>{button}</div>
      <div
        className={className(`${prefixCls}-closeIcon`)}
        style={{ display: closeable ? 'block' : 'none' }}
        onClick={onCloseBanner}
        aria-hidden="true"
      >
        <Close />
      </div>
    </div>
  );
};

export default Banner;
