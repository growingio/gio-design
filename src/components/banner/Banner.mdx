import { Subtitle, ArgsTable, Story } from '@storybook/addon-docs/blocks';
import Banner from './Banner';

# Banner 横幅

<Subtitle>将重要的信息通知到用户，例如：线下活动报名，系统维护等通知</Subtitle>

## 参数说明

<ArgsTable of={Banner} />

## 代码演示

### 信息通知

<Story id="functional-components-banner--base" />

### 警告通知

<Story id="functional-components-banner--alert" />

### 可关闭

<Story id="functional-components-banner--closeable" />
