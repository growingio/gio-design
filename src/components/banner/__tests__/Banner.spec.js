import React from 'react';
import '../../../../es/components/banner/style/index.css';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Banner from '../Banner';

describe('Testing Banner', () => {
  it('should be stable', () => {
    const wrapper = renderer.create(<Banner>快照</Banner>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(<Banner />);
      wrapper.setProps({
        type: 'normal',
        closeable: true,
        content: 'content',
      });
      wrapper.unmount();
    }).not.toThrow();
  });

  it('props content', () => {
    const stringwrapper = mount(<Banner type="normal" content="content" />).children();
    expect(stringwrapper.childAt(0).text()).toBe('content');
    const nodeWrapper = mount(<Banner content={<div>content</div>} />).children();
    expect(nodeWrapper.childAt(0).type()).toBe('div');
  });

  it('props type', () => {
    expect(mount(<Banner type="normal" />).exists('.gio-banner-normal')).toBe(true);
    expect(mount(<Banner type="alert" />).exists('.gio-banner-alert')).toBe(true);
  });

  it('can be click to close', () => {
    const closeMock = jest.fn();
    const wrapper = mount(<Banner onClose={closeMock}>关闭</Banner>);
    wrapper.find('.gio-banner-closeIcon').simulate('click');
    expect(closeMock).toBeCalled();
  });
});
