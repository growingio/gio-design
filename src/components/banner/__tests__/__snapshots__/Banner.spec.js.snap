// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Banner should be stable 1`] = `
<div
  className="gio-banner gio-banner-normal gio-banner-closeable"
>
  <div
    className="gio-banner-content"
  />
  <div
    className="gio-banner-button"
  />
  <div
    aria-hidden="true"
    className="gio-banner-closeIcon"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
      }
    }
  >
    <span
      aria-label="close-outlined"
      className="gio-icon"
      role="img"
    >
      <svg
        fill="currentColor"
        height="16px"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 14 14"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <style />
        </defs>
        <g>
          <g>
            <path
              d="M6.62 5.92a.54.54 0 00.76 0l3.87-3.87a.49.49 0 01.7.7L8.08 6.62a.54.54 0 000 .76L12 11.25a.49.49 0 01-.7.7L7.38 8.08a.54.54 0 00-.76 0L2.75 12a.49.49 0 01-.7-.7l3.87-3.92a.54.54 0 000-.76L2.05 2.75a.49.49 0 01.7-.7z"
              fill="currentColor"
              fillRule="evenodd"
              id="close-outlined_svg__close"
            />
          </g>
        </g>
      </svg>
    </span>
  </div>
</div>
`;
